File: 1049/main.py
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        sum = 0
        for x in stones:
            sum += x
        m = sum//2
        n = len(stones)

        dp = [[0 for _ in range(m + 1)] for _ in range(n + 1)]

        for i in range(0, n + 1):
            for j in range(0, m + 1):                
                if i == 0 and j == 0:
                    dp[i][j] = 1
                elif i == 0:
                    dp[i][j] = 0
                elif j == 0:
                    dp[i][j] = 1
                elif stones[i - 1] <= j:
                    dp[i][j] = dp[i - 1][j] | dp[i - 1][j - stones[i - 1]]
                else:
                    dp[i][j] = dp[i - 1][j]

        ans = 0
        for j in range(m, -1, -1):
            if dp[n][j] != 0:
                ans = j
                break
                
        return sum - 2 * ans;

File: 1049/main.cpp
class Solution {
public:
	int lastStoneWeightII(vector<int>& stones) {
		// + + - - + - + -
		int sum = 0;
		for (int x : stones)
			sum += x;
		int m = sum / 2;
		int n = stones.size();

		vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
		for (int i = 0; i <= n; i++) {
			for (int j = 0; j <= m; j++) {
				if (i == 0 and j == 0)
					dp[i][j] = 1;
				else if (i == 0)
					dp[i][j] = 0;
				else if (j == 0)
					dp[i][j] = 1;
				else if (stones[i - 1] <= j)
					dp[i][j] = dp[i - 1][j - stones[i - 1]] || dp[i - 1][j];
				else
					dp[i][j] = dp[i - 1][j];
			}
		}

		int ans = 0;
		for (int j = m; j >= 0; j--) {
			if (dp[n][j]) {
				ans = j;
				break;
			}
		}

		return sum - 2 * ans;
	}
};